modules/navier_stokes/src/materials/GeneralFunctorFluidProps.C

```diff
@@ -46,14 +50,15 @@ GeneralFunctorFluidProps::GeneralFunctorFluidProps(const InputParameters & param
     _pressure(getFunctor<ADReal>(NS::pressure)),
     _T_fluid(getFunctor<ADReal>(NS::T_fluid)),
     _speed(getFunctor<ADReal>(NS::speed)),
+    _force_define_density(getParam<bool>("force_define_density")),
     _rho(getFunctor<ADReal>(NS::density)),
     _mu_rampdown(getFunction("mu_rampdown"))
 {
   //
   // Set material properties functors
   //
 
-  if (!isParamValid(NS::density))
+  if (!isParamValid(NS::density) || _force_define_density)
```

Comment:
Isn't this problematic for our unique functor naming?

Reply:
only if you did define density

the problem arises with
[GlobalParams]
  rho = 'rho'
[]

which makes GeneralFunctorFluidProps think you are providing it with a functor parameter.
There are a few other solutions if you dont like this:
- rename density parameter to GFFP so that GlobalParams dont hit
- dont check for isParamValid but isFunctorExistsAlready or is rho_param a constant functor?

Reply:
What if there were two objects, `GeneralFunctorIncompressibleFluidProps` and `GeneralFunctorWeaklyCompressibleFluidProps` with the latter inheriting from the former. The former would not define density but the latter would. Would that cover all of our use cases? I guess in your other comment you're suggesting that even for an incompressible simulation you may want to create density from the `GeneralFunctor*FluidProps`, leveraging `SimpleFluidProperties`?

Reply:
I m not sure there are enough differences between the two to warrant two classes. Especially since the incompressible one is not likely to be used much, since it's only for someone who would go through SFP. It's definitely a clean solution code-wise. 

leverage `virtual` for what purpose?

Reply:
> leverage virtual for what purpose?

not sure exactly, but like in INSFV vs PINSFV in some places I have a `virtual` epsilon method that returns a functor which is unity for INSFV but an actual porosity functor for PINSFV. And Pronghorn historically has loved doing `virtuals` in its materials

Reply:
> I m not sure there are enough differences between the two to warrant two classes

Ok, yea I'm fine with this as it is. If someone complains later we can consider designing differently